[WindowSystem,Info] Add Display HDR capabilities to system info Video category

quietvoid opened the pull request:
## Description
This change allows Kodi to know what the display supports as HDR formats, through Android APIs.
Other platforms work as long as `IsHDRDisplay()` is implemented for them. Then only `HDR10` is assumed supported.

## Motivation and context
This is useful information to know.
Also it will help maintain backwards compatibility in the future as we can detect whether the playback chain supports a codec fully.

## How has this been tested?
Tested on a FireTV Stick 4K Max.

## What is the effect on users?
Adds a line to the System info GUI and logging on window initialization.

Example log:
```
DEBUG <general>: CAndroidUtils: Display Supported HDR types: Dolby Vision HDR10 HLG
```

## Screenshots (if appropriate):
When HDR is supported (LG C8):
![screen-hdr](https://user-images.githubusercontent.com/39477805/139595690-5b1c18bb-44b3-4f46-9129-a4547c1f1f26.png)

When disabling HDR through the device settings:
![screen-sdr](https://user-images.githubusercontent.com/39477805/139520565-a9997cb6-2f26-440e-84cf-b09978c68dde.png)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

thexai commented:
In general, it seems like a very good idea, although there are some details that will have to be polished.

thexai commented on xbmc/guilib/guiinfo/SystemGUIInfo.cpp:
```diff
@@ -321,6 +321,48 @@ bool CSystemGUIInfo::GetLabel(std::string& value, const CFileItem *item, int con
       return true;
 #endif
 
+#if defined(TARGET_ANDROID)
```
HDR is supported on Linux and Windows too. There is not reason to show these infos only on Android. 

On platforms that not support any HDR type "none" informative string is fine too.

`IsHDRDisplay()` can be called on all platforms.

quietvoid replied:
Yea I'm aware, but there's no implementations for the new methods.
So for these platforms it would be assumed HDR10, as far as I know.

thexai replied:
Don't worry: as `DisplaySupportsDolbyVision()` is defined on WinSystem.h and returns false by default, on all platforms where is not supported or is not implemented yet, `DisplaySupportsDolbyVision` can be called anyway and returns false.

quietvoid replied:
Yes, if `IsHDRDisplay()` is false then the System Info already displays `None` for the HDR support.

thexai commented on xbmc/windowing/android/AndroidUtils.h:
```diff
@@ -31,7 +31,18 @@ class CAndroidUtils : public ISettingCallback
   void OnSettingChanged(const std::shared_ptr<const CSetting>& setting) override;
 
+  // HDR info
+  static const int HDR_TYPE_DOLBY_VISION = 1;
```
Maybe convert to enum or enum class

quietvoid replied:
`enum class` did not work for the `find` in the other methods:
`std::find(types.begin(), types.end(), CAndroidUtils::HDRTypes::HLG) != types.end();`

I didn't bother keeping it and just changed to `enum`

CastagnaIT commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -348,3 +376,36 @@ void CAndroidUtils::OnSettingChanged(const std::shared_ptr<const CSetting>& sett
+
+  for (int type : hdrTypes)
+  {
+    CLog::Log(LOGDEBUG, "CAndroidUtils: Display Supported HDR type: {}", HdrTypeString(type));
```
i suggest to collect the strings to print in a single debug line instead to print a new log line for each type, the log is already heavy

quietvoid replied:
Thanks, done.
Now only one line at init is printed: `DEBUG <general>: CAndroidUtils: Display Supported HDR types: Dolby Vision HDR10 HLG `

ksooo commented on xbmc/windowing/android/AndroidUtils.h:
```diff
@@ -31,7 +31,18 @@ class CAndroidUtils : public ISettingCallback
+  static const int HDR_TYPE_HLG = 3;
+  static const int HDR_TYPE_HDR10_PLUS = 4;
+
+  std::vector<int> GetDisplaySupportedHdrTypes();
```
could this member function be `const`?

ksooo commented on xbmc/windowing/android/AndroidUtils.h:
```diff
@@ -31,7 +31,18 @@ class CAndroidUtils : public ISettingCallback
   mutable int m_height;
+
+private:
+  void LogDisplaySupportedHdrTypes();
```
could this member function be `const`?

ksooo commented on xbmc/windowing/android/WinSystemAndroid.cpp:
```diff
@@ -324,3 +324,27 @@ std::unique_ptr<WINDOWING::IOSScreenSaver> CWinSystemAndroid::GetOSScreenSaverIm
+
+bool CWinSystemAndroid::DisplaySupportsHDR10()
+{
+  std::vector<int> types = m_android->GetDisplaySupportedHdrTypes();
```
`const std::vector...`

thexai commented:
> It seams this PR does not consider multiple displays connected to the same device? So, to which of my attached two screens the info refers?

It is enough that the information corresponds to the screen where the Kodi window is at that moment. In Windows it works like this.

ksooo commented on xbmc/windowing/android/WinSystemAndroid.cpp:
```diff
@@ -324,3 +324,27 @@ std::unique_ptr<WINDOWING::IOSScreenSaver> CWinSystemAndroid::GetOSScreenSaverIm
+
+bool CWinSystemAndroid::DisplaySupportsHLG()
+{
+  std::vector<int> types = m_android->GetDisplaySupportedHdrTypes();
```
`const std::vector...`

ksooo commented on xbmc/windowing/android/WinSystemAndroid.cpp:
```diff
@@ -324,3 +324,27 @@ std::unique_ptr<WINDOWING::IOSScreenSaver> CWinSystemAndroid::GetOSScreenSaverIm
+
+bool CWinSystemAndroid::DisplaySupportsHDR10Plus()
+{
+  std::vector<int> types = m_android->GetDisplaySupportedHdrTypes();
```
`const std::vector...`

ksooo commented on xbmc/windowing/android/WinSystemAndroid.cpp:
```diff
@@ -324,3 +324,27 @@ std::unique_ptr<WINDOWING::IOSScreenSaver> CWinSystemAndroid::GetOSScreenSaverIm
+
+bool CWinSystemAndroid::DisplaySupportsDolbyVision()
+{
+  std::vector<int> types = m_android->GetDisplaySupportedHdrTypes();
```
`const std::vector...`

ksooo commented on xbmc/windows/GUIWindowSystemInfo.cpp:
```diff
@@ -161,6 +161,10 @@ void CGUIWindowSystemInfo::FrameMove()
 
     SetControlLabel(i++, "{} {}", 22010, SYSTEM_GPU_TEMPERATURE);
+
+#if defined(TARGET_ANDROID)
```
I would prefer if this feature would need no platform "ifdeffery", backing what thexai said.

ksooo reviewed (commented):
Nice improvement, needs a bit more love though.

quietvoid commented on xbmc/windowing/WinSystem.h:
```diff
@@ -173,6 +173,11 @@ class CWinSystemBase
 
   virtual std::vector<std::string> GetConnectedOutputs() { return {}; }
 
+  virtual bool DisplaySupportsHDR10() { return false; }
```
Ok.

quietvoid replied:
I commented on the fix for `IsHDRDisplay()` on Android.

quietvoid commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -123,6 +123,31 @@ static void fetchDisplayModes()
+    case CAndroidUtils::HDR_TYPE_HDR10_PLUS:
+      value = "HDR10+";
+      break;
+    case -1:
```
I don't even think the `-1` is related to the `HdrCapabilities`
https://developer.android.com/reference/android/view/Display.HdrCapabilities

I might just remove, or make a default case with Unknown or something.

ksooo commented:
I thought I can use screen 1 for Kodi GUI and another screen to playback video. So Kodi can be at two screens „at the moment“.

thexai commented:
It is not an issue that has to do with this PR: android already has an `IsHDRdisplay` method, the only thing this PR does is show this information also in SystemInfo and add new methods that do the same or to DV, HLG, etc. Your concern about what happens when there is more than one display is fine but it is not a new problem that this PR should solve.

jjd-uk commented:
> I thought I can use screen 1 for Kodi GUI and another screen to playback video. So Kodi can be at two screens „at the moment“.

I didn't think that was possible, or is that a Linux only thing???

quietvoid commented on xbmc/guilib/guiinfo/SystemGUIInfo.cpp:
```diff
@@ -321,6 +321,48 @@ bool CSystemGUIInfo::GetLabel(std::string& value, const CFileItem *item, int con
+        if (CServiceBroker::GetWinSystem()->DisplaySupportsHDR10())
+          hdr10 = localizedYes;
+        else
+          hdr10 = localizedNo;
```
ternary operators do however make for long lines.
I'll probably let Jenkins give me a diff.

quietvoid commented:
I have replaced `CWinSystemAndroidGLESContext::m_hasEGLHDRExtensions` with member variables for each of the extension.
The existing behaviour was kept the same, but now `isHdrDisplay()` is less strict and returns true if at least ST2086 (HDR10) metadata extension is available.

ksooo commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -123,6 +124,26 @@ static void fetchDisplayModes()
+      case CAndroidUtils::HDRTypes::HDR10_PLUS:
+        return "HDR10+";
+      default:
+        return "Unknown";
```
This should be localized.

thexai commented:
> I have not yet tested on other platforms like Windows because my environment isn't setup for that.

I think everything is fine for Windows now although I will runtime it sometime.

ksooo commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -123,6 +124,26 @@ static void fetchDisplayModes()
 
+namespace
+{
+  std::string HdrTypeString(int type)
```
Wrong indentation

quietvoid replied:
Yes I'm applying Jenkins' fixes in the next push.

quietvoid commented on xbmc/windowing/android/AndroidUtils.h:
```diff
@@ -31,7 +31,20 @@ class CAndroidUtils : public ISettingCallback
   void OnSettingChanged(const std::shared_ptr<const CSetting>& setting) override;
 
+  // HDR info
+  enum HDRTypes {
```
Fixed.

quietvoid commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -192,6 +213,8 @@ CAndroidUtils::CAndroidUtils()
     CAndroidUtils::SETTING_LIMITGUI
   });
+
+  CAndroidUtils::LogDisplaySupportedHdrTypes();
```
Thanks, corrected.

ksooo commented:
> It is not an issue that has to do with this PR: android already has an IsHDRdisplay method

@thexai, IsHDRDisplay() is only used in VideoPlayer context, means it is only called when the right display (the display showing the playing video) is active. Thus, no problem with the current use cases.

 This PR uses this and the new methods in the UI context, which could be the wrong display. So, strictly speaking this PR introduces a bug.
 
  No hard feelings, and wont discuss this to death, but why can't we just do it right?

thexai commented:
> IsHDRDisplay() is only used in VideoPlayer context, means it is only called when the right display (the display showing the playing video) is active. 

Wrong. Also is used to determine if optional setting is visible or not:

![Captura de pantalla 2021-10-30 192552](https://user-images.githubusercontent.com/58434170/139543077-7b264cd9-db3c-4d33-95ed-816837bc28d8.png)

ksooo commented:
That’s bad. But does this legitimate to manifest this design flaw even more by introducing more of this?

thexai commented:
I do not want to talk without being completely sure and trying it but I think there is no defect in Windows: if there are two monitors and one is HDR and the other is not, the value that is taken corresponds to the one in the where Kodi (active window) is at that moment.

You can say then that window can be between two screens.... this is possibly an unforeseen scenario. Also you can unplug HDR monitor and plug another SDR monitor while Kodi is running and maybe something goes wrong.... 😄

thexai commented:
Runtime tested now:  all is fine on Windows.

No "design flaws". HDR infos changes and updates  **in real-time** when the Kodi Window is more than 50% into HDR screen:

**100% on SDR screen:**
![screenshot00000](https://user-images.githubusercontent.com/58434170/139554120-eb8d723f-7bfd-4a34-b0da-b453a80a1a0b.png)

**more than 50% on HDR screen:**
![Sin título](https://user-images.githubusercontent.com/58434170/139554139-8bf42b95-c794-4340-aad5-2b065f1ecce8.png)

thexai reviewed (approved):
Thank you very much for the improvement.

As is a change that affect all platforms, is need at least two team members approvals to merge.

CastagnaIT commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -348,3 +372,38 @@ void CAndroidUtils::OnSettingChanged(const std::shared_ptr<const CSetting>& sett
+    ss << HdrTypeString(type) << " ";
+  }
+
+  CLog::Log(LOGDEBUG, "CAndroidUtils: Display supported HDR types: {}", ss.str());
```
here you are working with strings, you can avoid to use stringstream and use a simple std::string

quietvoid replied:
That's true. Initially I was trying to make it prettier with a separator.

lrusak commented on xbmc/windowing/WinSystem.h:
```diff
@@ -173,6 +173,10 @@ class CWinSystemBase
 
   virtual std::vector<std::string> GetConnectedOutputs() { return {}; }
 
+  virtual bool DisplaySupportsHLG() { return false; }
```
CWinSystem seems to be getting more and more bloated by HDR stuff.is it really necessary to add three methods purely for indication? Do you think we could better handle thing in our current HDR handling?

thexai replied:
> is it really necessary to add three methods purely for indication?

I think the idea behind quietvoid is to use these methods in the future to handle DV streams in the right way. Some extensions / codec overrides need to be added but only on systems that support DV. Otherwise fallback HDR10 compatibility is broken. That is, some streams are HDR10 + DV compatible and the codec to be used is selected based on the capabilities of system (or the current device/firmware settings).

I also suggested something like this in another PR: https://github.com/xbmc/xbmc/pull/19099#issuecomment-950202096

> Do you think we could better handle thing in our current HDR handling?

Indeed, there are many different ways to implement this. For example we could use a single method and return a structure with the supported HDR types. But this required touching the code that already exists with the risk of introducing errors. Since this is a "work in progress" I am fine with introducing new methods on the fly. Perhaps later it will be seen more clearly if it can be re-organized / unified in a better way.

Of course this is just my opinion. @irusak, I know that you are or were working on HDR Linux / LibreELEC things so I wanted you to participate in this PR and you can see things that I do not see or that I am not even aware of.

quietvoid replied:
The idea here is simply to have access to the HDR capabilities globally, and I just ended up on `CWinSystem` as a convenient class as it already used `CAndroidUtils::IsHDRDisplay()`.

An alternative could just be a list of a new enum type for all the common HDR formats, and some helper method to avoid boilerplate everywhere.

quietvoid replied:
I created a class as an alternative to the different functions in `CWinSystem`.
Do let me know if it's a viable improvement over the previous idea.

lrusak commented on xbmc/windowing/android/WinSystemAndroidGLESContext.h:
```diff
@@ -41,6 +41,10 @@ class CWinSystemAndroidGLESContext : public CWinSystemAndroid, public CRenderSys
   bool IsHDRDisplay() override;
   bool SetHDR(const VideoPicture* videoPicture) override;
 
+  bool DisplaySupportsHLG() override;
```
Pending my other comment.

These overrides shouldn't be needed as this class derives CWinSystemAndroid which derives CWinSystem.

quietvoid replied:
Indeed they're unnecessary, I've removed them.

CastagnaIT commented:
I was wondering why forcing to display of each type of capabilities in the GUI info?
`Display supported HDR types` should show only the supported one without all these `yes` or `no` affermations, for me this has no much sense

enen92 commented:
If I may, I also support the request above. I'd prefer to have a list of supported modes (and none if none is supported) instead of listing them all with yes/no.

enen92 commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -1728,6 +1728,12 @@ const infomap weather[] =        {{ "isfetched",        WEATHER_IS_FETCHED },
+///   \table_row3{   <b>`System.SupportedHDRTypes`</b>,
+///                  \anchor System_SupportedHDRTypes
+///                  _string_,
+///     @return The display's supported HDR types.
```
misses skinning api revision like the others above:
```
///     <p><hr>
///     @skinning_v20 **[New Infolabel]** \link  System_SupportedHDRTypes `
```

quietvoid replied:
Thanks.

CastagnaIT commented on xbmc/windowing/android/AndroidUtils.h:
```diff
@@ -31,7 +31,21 @@ class CAndroidUtils : public ISettingCallback
+    HDR10 = 2,
+    HLG = 3,
+    HDR10_PLUS = 4
+  };
```
in an enum is not needed assign the value on each item only in the first one to set the starting value, unless you intend to do something specific in the future

CastagnaIT replied:
now understand, this is a mapping of https://developer.android.com/reference/android/view/Display.HdrCapabilities#constants_1
then maybe here is better add a comment on this enum that this refer to these android constants?

quietvoid replied:
I thought the fact that it's in AndroidUtils was implying that, but I'll add a comment.

CastagnaIT commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -123,6 +124,26 @@ static void fetchDisplayModes()
+      return g_localizeStrings.Get(13205); // Unknown
+  }
+}
+} // namespace
```
`// namespace` --> `// unnamed namespace`

CastagnaIT commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -348,3 +371,38 @@ void CAndroidUtils::OnSettingChanged(const std::shared_ptr<const CSetting>& sett
+    text += HdrTypeString(type) + " ";
+  }
+
+  CLog::Log(LOGDEBUG, "CAndroidUtils: Display supported HDR types: {}", text);
```
there is potential for improvement here by moving the space to the beginning:
`text += " " + HdrTypeString(type);`
and:
`CLog::Log(LOGDEBUG, "CAndroidUtils: Display supported HDR types:{}", text);`
then the space at the end is now removed

CastagnaIT replied:
can you explain what happens when nothing is supported? is an empty string?

quietvoid replied:
> can you explain what happens when nothing is supported? is an empty string?

Yes that's what happens.

CastagnaIT replied:
i think for a better understanding of the log is better add a ternary op. instead pass directly "text"var where when the "text" is empty then pass " None" string

quietvoid replied:
I'm not sure why but I can't get a localized string at the moment of the log. It's always empty.

CastagnaIT replied:
usually in the log we should keep english language, not localized language, you can add directly " None"

howie-f commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -348,3 +371,38 @@ void CAndroidUtils::OnSettingChanged(const std::shared_ptr<const CSetting>& sett
     CDisplaySettings::GetInstance().ClearCalibrations();
 }
+
+const std::vector<int> CAndroidUtils::GetDisplaySupportedHdrTypes()
```
please never return plain const from functions.. either `const T&` or just `T` by value like in this case.

thexai replied:
I guess is just a confusion by https://github.com/xbmc/xbmc/pull/20422#discussion_r739639288

What @ksooo wanted is:
`std::vector<int> CAndroidUtils::GetDisplaySupportedHdrTypes() const`

howie-f replied:
yes right :) sorry i didn't see this comment. this ^^ is what we want

quietvoid replied:
Okay, I got confused with the header being `const`.

thexai replied:
I know the review process is stressful sometimes... but we are like that ;-)

quietvoid replied:
No problem, this should be fixed now.

howie-f commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -348,3 +371,38 @@ void CAndroidUtils::OnSettingChanged(const std::shared_ptr<const CSetting>& sett
+  return {};
+}
+
+const void CAndroidUtils::LogDisplaySupportedHdrTypes()
```
just `void` please

thexai replied:
same:
`void CAndroidUtils::LogDisplaySupportedHdrTypes() const`

ksooo replied:
what thexai said was the intention of my comment "Could this member function be const?".

quietvoid commented:
Corrected the System info to be the list of supported types. Updated the screenshots in the description.
Still have to figure out a better design if we don't want a function for every type in `CWinSystem`.

thexai commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -348,3 +370,39 @@ void CAndroidUtils::OnSettingChanged(const std::shared_ptr<const CSetting>& sett
+  const std::vector<int> hdrTypes = GetDisplaySupportedHdrTypes();
+  std::string text;
+
+  for (int type : hdrTypes)
```
```suggestion
  for (const auto& type : hdrTypes)
```

ksooo replied:
what's wrong with int here? Are we going to replace all types, even simplest ones with `auto`? ;-)

thexai replied:
`const int&`  then, but I prefer auto in such cases :)

CastagnaIT replied:
i agree with @ksooo "auto" is not really needed i could understand if it was an iterator or some else of long...

thexai replied:
![Captura de pantalla 2021-10-31 201030](https://user-images.githubusercontent.com/58434170/139597951-d4d13568-9009-4811-a36e-9b74cec3cf6e.png)

ksooo replied:
> const int& then, but I prefer auto in such cases :)

 As you say it's all about personal preference, there is no right and wrong here. And we don't have a coding guideline for this, iirc. I alos use const auto& heavily, but not for trivial types and never if the code gets longer due to this. This contradicts the idea of auto.
 
  Let the submitter of this PR decide on this.

thexai replied:
Seems use `const &` is more efficient:
https://github.com/xbmc/xbmc/pull/18667

I said it for this, rather than for auto/int

quietvoid replied:
Will change to `const int&`.

CastagnaIT commented:
Under GUI/Log output prospective for me now is good, no further requests from me
thank you

quietvoid commented:
I made a `CHDRCapabilities` class for windowing systems to return, it reduces the added stuff in `CWinSystem`.
There's most likely something wrong with how I'm doing it, though.

I couldn't replace the `IsHdrDisplay` function because it's used in settings and I don't think my change is an alternative for it.

lrusak commented on xbmc/windowing/android/AndroidUtils.cpp:
```diff
@@ -123,6 +123,26 @@ static void fetchDisplayModes()
+{
+std::string HdrTypeString(int type)
+{
+  switch (type)
```
Maybe a std::map is better here?

quietvoid replied:
I'm not sure if it would be better, because the types are coming from Android and there's always a possibility of a new one being added.
A map wouldn't have a default fallback in that case, and would require extra logic.

lrusak replied:
```
const std::map<int, const char*> hdrMap =
{
...
}

auto hdr = hdrMap.find(type);
if (hdr != hdrMap.end())
  return hdr.second;

return "unknown"
```

Done on my phone so it's paraphrased. The map could probably go in an anonymous namespace and this method probably wont be needed.

Just my 2c.

quietvoid replied:
Okay, that makes sense.
I'll change it. Thanks.

lrusak commented on xbmc/utils/HDRCapabilities.h:
```diff
@@ -0,0 +1,32 @@
+  bool SupportsHDR10Plus() const { return m_hdr10_plus; }
+  bool SupportsDolbyVision() const { return m_dolby_vision; }
+
+  void SetHDR10(bool supported) { m_hdr10 = supported; }
```
Probably don't need the bool param for these methods as they will only ever be set to true.

quietvoid replied:
Indeed.

thexai commented:
@lrusak , do you agree with the latest changes?

I would like to merge this.

enen92 commented on xbmc/guilib/guiinfo/SystemGUIInfo.cpp:
```diff
@@ -320,6 +320,32 @@ bool CSystemGUIInfo::GetLabel(std::string& value, const CFileItem *item, int con
+      }
+      else
+      {
+        // None
```
Hopefully I am not late to the party but I think this can be improved. Infolabels are resources that are exposed to multiple subsystems: GUI, skins, python, etc. They should contain values that can be easily processed.
If you make this dependent on the user Kodi configured language, you break this rule - i.e. it's hard to know, e.g. from python, that no HDR mode is supported.
I'd suggest to return an empty string here and translate that in GUIWindowSystemInfo using a ternary operator or something similar (depending on whether the infolabel isEmpty())

quietvoid replied:
Should be fixed now.
Thank you.

quietvoid committed:
windowing/WinSystem: Add methods to access display HDR capabilities

New class for generic HDR formats.
Implemented for WinSystemAndroid only for now.

Signed-off-by: quietvoid <39477805+quietvoid@users.noreply.github.com>

quietvoid committed:
GUIWindowSystemInfo: Add display HDR capabilities in Video system info

Implemented fully for Android, other platforms will only show HDR10.

Signed-off-by: quietvoid <39477805+quietvoid@users.noreply.github.com>

enen92 reviewed (approved):
Lgtm, thanks

matthuisman commented:
Will this info be available for Kodi addons? 
Looks like it may just be set as an info label?
So maybe an addon could use the built-in to get the info label and then look for what it needs in the returned string?

Add-ons like Disney+ send the device capabilities when requesting a manifest. Currently need a toggle in the settings. Would be nice to automate it.

On another note, say a manifest has dolby vision and hdr10, will Kodi now select the hdr10 if the display doesn't support dolby vision? Or is this purely cosmetic currently?

quietvoid commented:
I don't personally know how the addon interfaces work, so not sure.

> Add-ons like Disney+ send the device capabilities when requesting a manifest. Currently need a toggle in the settings. Would be nice to automate it.
Is this similar to how the Netflix addon works? I do think leaving the option to the user is better than automating according to the max capabilities.
Maybe both can be possible, with a block list for the user preference.

> On another note, say a manifest has dolby vision and hdr10, will Kodi now select the hdr10 if the display doesn't support dolby vision? Or is this purely cosmetic currently?
Currently only cosmetic. It's probably possible to do what you want but I wouldn't know where to start.

CastagnaIT commented:
> Will this info be available for Kodi addons?
> Looks like it may just be set as an info label?

@matthuisman i think you should test this PR 19983 that implement the new mediaflag `HdrType`:
https://github.com/xbmc/xbmc/pull/19983/files#diff-4a9e9842cc1d48d4cb20b2c78ccb3a3c89315d865ef8c599bd21c4fcea1d1794R3703-R3712

> Add-ons like Disney+ send the device capabilities when requesting a manifest. Currently need a toggle in the settings. Would be nice to automate it.

yeah It's something i thought to add about last year,
add the possibility to addons to know if a system support HDR and if it is enabled via python API,
not so difficult to add, but as discussed in the past i remember that devs wanted to introduce a new python module like `xbmcsystem` where add this features and others

if no one will do it i can think to implement this new python module when i will finish developing/fixing the subtitles source code...not immediate thing

quietvoid commented:
Thought I tested the Firestick 4K 2018 but apparently not.
```
2021-11-19 18:10:07.155 T:6236     INFO <general>: failed to find EGL configs with appropriate attributes (EGL_BAD_ATTRIBUTE)
2021-11-19 18:10:07.156 T:6236    DEBUG <general>: CWinSystemAndroidGLESContext::InitWindowSystem: HDRConfig: 0, HDRExtensions: 0
...
CAndroidUtils: Display supported HDR types: Dolby Vision HDR10 HLG
```

So I guess `CWinSystemAndroidGLESContext::IsHDRDisplay()` is still broken on that one.

chewitt commented:
Is there a plan to support this under Linux? - If not, perhaps we could define this for Android only as the line item shows on Linux/ARM images (LibreELEC) where my TV supports (and Kodi plays) HDR content but the GUI shows "none" :(

thexai commented:
In truth, this does not depend on this PR.

For a platform (Linux) to report that it supports HDR, it must implement at least this method:

https://github.com/xbmc/xbmc/blob/72f47fab7b932fd749aad392ba2de10935429a0b/xbmc/windowing/WinSystem.h#L166

(HDR screen detection)
